storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,766 @@
+        }
+    }
+    
+    static class DynamicState {
+        public final MachineState state;
+        public final LocalAssignment newAssignment;
+        public final LocalAssignment currentAssignment;
+        public final Container container;
+        public final LocalAssignment pendingLocalization;
+        public final Future<Void> pendingDownload;
+        public final Set<TopoProfileAction> profileActions;
+        public final Set<TopoProfileAction> pendingStopProfileActions;
+        
+        /**
+         * The last time that WAITING_FOR_WORKER_START, KILL, or KILL_AND_RELAUNCH were entered into.
+         */
+        public final long startTime;
+        
+        public DynamicState(final LocalAssignment currentAssignment, Container container, final LocalAssignment newAssignment) {
+            this.currentAssignment = currentAssignment;
+            this.container = container;
+            if ((currentAssignment == null) ^ (container == null)) {
+                throw new IllegalArgumentException("Container and current assignment must both be null, or neither can be null");
+            }
+            
+            if (currentAssignment == null) {
+                state = MachineState.EMPTY;
+            } else {
+                state = MachineState.RUNNING;
+            }
+            
+            this.startTime = System.currentTimeMillis();
+            this.newAssignment = newAssignment;
+            this.pendingLocalization = null;
+            this.pendingDownload = null;
+            this.profileActions = new HashSet<>();
+            this.pendingStopProfileActions = new HashSet<>();
+        }
+        
+        public DynamicState(final MachineState state, final LocalAssignment newAssignment,
+                final Container container, final LocalAssignment currentAssignment,
+                final LocalAssignment pendingLocalization, final long startTime,
+                final Future<Void> pendingDownload, final Set<TopoProfileAction> profileActions, 
+                final Set<TopoProfileAction> pendingStopProfileActions) {
+            this.state = state;
+            this.newAssignment = newAssignment;
+            this.currentAssignment = currentAssignment;
+            this.container = container;
+            this.pendingLocalization = pendingLocalization;
+            this.startTime = startTime;
+            this.pendingDownload = pendingDownload;
+            this.profileActions = profileActions;
+            this.pendingStopProfileActions = pendingStopProfileActions;
+        }
+        
+        public String toString() {
+            StringBuffer sb = new StringBuffer();
+            sb.append(state);
+            if (state == MachineState.WAITING_FOR_WORKER_START ||
+                state == MachineState.KILL ||
+                state == MachineState.KILL_AND_RELAUNCH) {
+                sb.append(" msInState: ");
+                sb.append(Time.currentTimeMillis() - startTime);
+            }
```

Comment:
Although we don't need to track start times for every state for functionality, it might be nice to keep track of start times if only to print them in log messages. This is a nice-to-have suggestion and not required now. But it might be easy to add.